akka-stream/src/main/scala/akka/stream/impl/fusing/Interpreter.scala

```diff
@@ -282,18 +339,62 @@ private[akka] class OneBoundedInterpreter(ops: Seq[Stage[_, _]], val forkLimit:
       null
     }
 
-    override def hold(): FreeDirective = {
-      if (currentOp.holding) throw new IllegalStateException("Cannot hold while already holding")
-      currentOp.holding = true
+    override def holdUpstream(): UpstreamDirective = {
+      currentOp.precedingWasPull = false
+      currentOp.holdingUpstream = true
+      exit()
+    }
+
+    override def holdUpstreamAndPush(elem: Any): UpstreamDirective = {
```

Comment:
They properly model things that you can do: ball coming from downstream, we’re already holding the upstream ball, as a result we hold downstream and pull from upstream. How else would you express this?